metadata-ingestion/src/datahub/ingestion/api/source.py

```diff
@@ -553,17 +558,111 @@ def _apply_workunit_processors(
     def get_workunits(self) -> Iterable[MetadataWorkUnit]:
         workunit_processors = self.get_workunit_processors()
         workunit_processors.append(AutoSystemMetadata(self.ctx).stamp)
-        return self._apply_workunit_processors(
-            workunit_processors, auto_workunit(self.get_workunits_internal())
+
+        final_workunits = self.get_workunits_internal()
+        merged_workunits = self._merge_workunits(
+            input_workunits=final_workunits,
+            is_enabled_method=self.is_profiling_enabled_internal,
+            stage=IngestionHighStage.PROFILING,
+            method=self.get_profiling_internal,
+            method_name_to_method_dict={
+                # Have to pass Source.METHOD here to avoid the subclass methods from being passed here
+                # This is a to ensure subclasses have implemented these methods
+                "is_profiling_enabled_internal": Source.is_profiling_enabled_internal,
+                "get_profiling_internal": Source.get_profiling_internal,
+            },
+        )
+
+        yield from self._apply_workunit_processors(
+            workunit_processors, auto_workunit(merged_workunits)
         )
 
+    def _are_methods_defined_in_subclass(
```

Comment:
uhh, this kinda hacky. 
I think it would be a better approach using Protocol:
```
  from typing import Protocol, runtime_checkable

  @runtime_checkable
  class ProfilingCapable(Protocol):
      def is_profiling_enabled_internal(self) -> bool: ...
      def get_profiling_internal(self) -> Iterable[Union[MetadataWorkUnit, MetadataChangeProposalWrapper, 
  Entity]]: ...
```

```
  # In Source class:
  def _merge_workunits(self, input_workunits, is_enabled_method, stage, method, method_name_to_method_dict):
      # Simpler check using isinstance
      if not isinstance(self, ProfilingCapable):
          return input_workunits
      if not is_enabled_method():
          return input_workunits
      # ... rest of method

```